### PR TITLE
#155931587 Create DetailActivity, retrofit and Access API

### PR DESCRIPTION
**Pull request by tomuhenry**
**What does this PR do?**
This PR handles DetailActivity file, retrofit, and accessing API

**Description of Task to be completed?**
- Create layout file for DetailActivity and launch it through MainActivity
- Implement retrofit
- Access the [Github API](https://api.github.com/search/users?q=location:Nairobi+language:Java)

**What are the relevant Pivotal Tracker stories?**
[#155820624](https://www.pivotaltracker.com/story/show/155820624)
[#155820645](https://www.pivotaltracker.com/story/show/155820645)
[#155810744](https://www.pivotaltracker.com/story/show/155810744) 
[#155810847](https://www.pivotaltracker.com/story/show/155810847)
[#155965831](https://www.pivotaltracker.com/story/show/155965831)
[#156004609](https://www.pivotaltracker.com/story/show/156004609)

Screen Shot 2019-04-10 at 2 13 47 PM

Screen Shot 2019-04-11 at 12 46 18 PM

Screen Shot 2019-04-11 at 12 46 38 PM

Screen Shot 2019-04-11 at 12 46 56 PM

### CODE REVIEW

**Comment by tomuhenry** [comment]
Hey @bibangamba, Could you please help with some reviews for this PR
Thanks.

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+
+        initViews();
+
+        swipeContainer = findViewById(R.id.swipeContainer);
+
+        //Configure refreshing color
+        swipeContainer.setColorSchemeResources(android.R.color.holo_orange_dark);
+        swipeContainer.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
```
Comment (bibangamba) [inline app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java]: please remove this implementation of the swipe to refresh layout as it is a deliverable you'll give in the UI/UX revamp later on

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                    List<Item> items = response.body().getItems();
+                    recyclerView.setAdapter(new ItemAdapter(items, getApplicationContext()));
+                    recyclerView.smoothScrollToPosition(0);
+                    swipeContainer.setRefreshing(false);
+                    pd.hide();
+                }
+
+                @Override
+                public void onFailure(Call<ItemResponse> call, Throwable t) {
+                    Log.d("Error", t.getMessage());
```
Comment (bibangamba) [inline app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java]: please remove all `Logs` as well as their import statements

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/ItemAdapter.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
+package com.andela.tomusange.javadevsnairobi.view;
+
+//import android.content.Context;
```
Comment (bibangamba) [inline app/src/main/java/com/andela/tomusange/javadevsnairobi/view/ItemAdapter.java]: please remove this whole block comment as you don't need it. please ensure that you remove all the other single line comments in your code. I've seen quite a number

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/DetailActivity.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        imageView = findViewById(R.id.user_image);
+        Username = findViewById(R.id.git_user_name);
+        Link = findViewById(R.id.github_url);
+
+        String username = getIntent().getExtras().getString("userName");
+        String avatarUrl = getIntent().getExtras().getString("imageUrl");
+        String link = getIntent().getExtras().getString("githubUrl");
+
+        Link.setText(link);
+        Linkify.addLinks(Link, Linkify.WEB_URLS);
```
Comment (bibangamba) [inline app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/DetailActivity.java]: very nice!!

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/model/ItemResponse.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package com.andela.tomusange.javadevsnairobi.view.model;
+
+import com.google.gson.annotations.Expose;
+import com.google.gson.annotations.SerializedName;
+
+import java.util.List;
+
+public class ItemResponse {
```
Comment (bibangamba) [inline app/src/main/java/com/andela/tomusange/javadevsnairobi/view/model/ItemResponse.java]: This name (`ItemResponse`) feels very generic. please change it to something that would make it easy for anyone to understand what the Class represents/is/does - [clean code](https://medium.com/mindorks/meaningful-names-a-dimension-of-writing-clean-code-fdae1ae4f0b1).

File: app/src/main/res/menu/detail.xml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (bibangamba) [inline app/src/main/res/menu/detail.xml:1]: please rename this file so it shows its meaning immediately. something like `share_menu` would have been easier to understand

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/model/Item.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
+package com.andela.tomusange.javadevsnairobi.view.model;
+
+import com.google.gson.annotations.Expose;
+import com.google.gson.annotations.SerializedName;
+
+public class Item {
```
Comment (bibangamba) [inline app/src/main/java/com/andela/tomusange/javadevsnairobi/view/model/Item.java]: please rename this class to show what it does/represents

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        pd.show();
+        recyclerView = findViewById(R.id.recycler_view);
+        recyclerView.setLayoutManager(new LinearLayoutManager(getApplicationContext()));
+        recyclerView.smoothScrollToPosition(0);
+        loadJSON();
+    }
+
+    private void loadJSON() {
+        Disconnected = findViewById(R.id.disconnected);
+        try{
```
Comment (bibangamba) [inline app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java]: this whole block should not be happening in the `MainActivity` class. you're not following the MVP pattern in this example. It should be moved into a presenter. this [link](https://github.com/jeancsanchez/Simple-MVP-Retrofit-example) shows an example of how you can partially implement the MVP architecture (although it is not perfect & will have to be improved when working on your D2 advancement outputs)

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+public class MainActivity extends AppCompatActivity {
+
+    private RecyclerView recyclerView;
+    TextView Disconnected;
+    ProgressDialog pd;
```
Comment (bibangamba) [inline app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java]: please use meaningful names in your code. `pd` is meaningless to anyone new to your code base. ensure to clean up most instances of this (`meaningful naming of variable, methods, classes, etc`) in your codebase

**Review by bibangamba — CHANGES_REQUESTED** [review]
the core implementation looks good @tomuhenry, but you have a lot of convention issues and code structuring issues. please fix them.

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/model/Item.java (hunk 7 above)
Comment (tomuhenry) [reply app/src/main/java/com/andela/tomusange/javadevsnairobi/view/model/Item.java]: Done

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java (hunk 8 above)
Comment (tomuhenry) [reply app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java]: Noted... Thanks. Gonna Refactor that.

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/DetailActivity.java (hunk 4 above)
Comment (tomuhenry) [reply app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/DetailActivity.java]: Thanks

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java (hunk 2 above)
Comment (tomuhenry) [reply app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java]: Done

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java (hunk 1 above)
Comment (tomuhenry) [reply app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java]: Ok... Removed

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java (hunk 9 above)
Comment (tomuhenry) [reply app/src/main/java/com/andela/tomusange/javadevsnairobi/view/controller/MainActivity.java]: Done

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/adapter/GithubUserAdapter.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+                    int position = getAdapterPosition();
+                    if (position != RecyclerView.NO_POSITION) {
+                        UserDetail clickedDataUserDetail = userDetails.get(position);
+                        Intent intent = new Intent(context, DetailActivity.class);
+                        intent.putExtra("userName", userDetails.get(position).getUserName());
+                        intent.putExtra("githubUrl", userDetails.get(position).getGithubUrl());
+                        intent.putExtra("imageUrl", userDetails.get(position).getImageUrl());
+                        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+                        context.startActivity(intent);
+                        Toast.makeText(v.getContext(), "You clicked " + clickedDataUserDetail.getUserName(), Toast.LENGTH_SHORT).show();
```
Comment (bibangamba) [inline app/src/main/java/com/andela/tomusange/javadevsnairobi/view/adapter/GithubUserAdapter.java:84]: please ensure to remove this in the next PR

File: app/src/main/res/menu/detail.xml (hunk 6 above)
Comment (bibangamba) [inline app/src/main/res/menu/detail.xml:1]: please rename this file in the next PR as well. `detail.xml` doesn't really convey the fact that it is a share menu. it doesn't convey anything really...

**Review by bibangamba — APPROVED** [review]
LGTM. I've left some comments on some minor (but important) things I'd like you to change. please look into changing them in the next PR since we are quite behind.

I'm going to merge this in.

File: app/src/main/java/com/andela/tomusange/javadevsnairobi/view/adapter/GithubUserAdapter.java (hunk 10 above)
Comment (tomuhenry) [reply app/src/main/java/com/andela/tomusange/javadevsnairobi/view/adapter/GithubUserAdapter.java:84]: I've removed the Toast

File: app/src/main/res/menu/detail.xml (hunk 6 above)
Comment (tomuhenry) [reply app/src/main/res/menu/detail.xml:1]: Renamed to `share_menu.xml`

**Pull request merged**